[mono][reflection] ArgumentNullException calling `GetCustomAttributes` on user-defined subtype of `ParameterInfo`

The following code works with .NET 8.0 CoreCLR, but throws with Mono:

`Program.cs`:

```csharp
using System.Reflection;

public class P {
    public static void Main()
    {
        var propi =typeof(P).GetProperty("Foo");
        var pi = new Retrofit.PropertyAsParameterInfo(propi!);
        IEnumerable<Attribute> cattrs = pi.GetCustomAttributes();
        Console.WriteLine($"Got {cattrs.Count()} attributes");
        foreach (var cattr in cattrs) {
            Console.WriteLine (cattr);
        }
    }

    [Flarg]
    [Blarg]
    public int Foo {get; set;}

    // public static void Foo([Flarg][Blarg]int i) {}
}

[AttributeUsage(AttributeTargets.Parameter | AttributeTargets.Property)]
public class FlargAttribute : Attribute {}

[AttributeUsage(AttributeTargets.Parameter | AttributeTargets.Property)]
public class BlargAttribute : Attribute {}
```

`PropertyAsParameterInfo.cs`:

```csharp
using System.Reflection;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;

namespace Retrofit;

internal sealed class PropertyAsParameterInfo : ParameterInfo
{
    private readonly PropertyInfo _underlyingProperty;
    private readonly ParameterInfo? _constructionParameterInfo;

    public PropertyAsParameterInfo(PropertyInfo propertyInfo)
    {
        Debug.Assert(propertyInfo != null, "PropertyInfo must be provided.");

        AttrsImpl = (ParameterAttributes)propertyInfo.Attributes;
        NameImpl = propertyInfo.Name;
        MemberImpl = propertyInfo;
        ClassImpl = propertyInfo.PropertyType;

        // It is not a real parameter in the delegate, so,
        // not defining a real position.
        PositionImpl = -1;

        _underlyingProperty = propertyInfo;
    }
    public override bool HasDefaultValue
        => _constructionParameterInfo is not null && _constructionParameterInfo.HasDefaultValue;
    public override object? DefaultValue
        => _constructionParameterInfo?.DefaultValue;
    public override int MetadataToken => _underlyingProperty.MetadataToken;
    public override object? RawDefaultValue
        => _constructionParameterInfo?.RawDefaultValue;

    public override object[] GetCustomAttributes(Type attributeType, bool inherit)
    {
        var constructorAttributes = _constructionParameterInfo?.GetCustomAttributes(attributeType, inherit);

        if (constructorAttributes == null || constructorAttributes is { Length: 0 })
        {
            return _underlyingProperty.GetCustomAttributes(attributeType, inherit);
        }

        var propertyAttributes = _underlyingProperty.GetCustomAttributes(attributeType, inherit);

        var mergedAttributes = new Attribute[constructorAttributes.Length + propertyAttributes.Length];
        Array.Copy(constructorAttributes, mergedAttributes, constructorAttributes.Length);
        Array.Copy(propertyAttributes, 0, mergedAttributes, constructorAttributes.Length, propertyAttributes.Length);

        return mergedAttributes;
    }

    public override object[] GetCustomAttributes(bool inherit)
    {
        var constructorAttributes = _constructionParameterInfo?.GetCustomAttributes(inherit);

        if (constructorAttributes == null || constructorAttributes is { Length: 0 })
        {
            return _underlyingProperty.GetCustomAttributes(inherit);
        }

        var propertyAttributes = _underlyingProperty.GetCustomAttributes(inherit);

        // Since the constructors attributes should take priority we will add them first,
        // as we usually call it as First() or FirstOrDefault() in the argument creation
        var mergedAttributes = new object[constructorAttributes.Length + propertyAttributes.Length];
        Array.Copy(constructorAttributes, mergedAttributes, constructorAttributes.Length);
        Array.Copy(propertyAttributes, 0, mergedAttributes, constructorAttributes.Length, propertyAttributes.Length);

        return mergedAttributes;
    }

    public override IList<CustomAttributeData> GetCustomAttributesData()
    {
        var attributes = new List<CustomAttributeData>(
            _constructionParameterInfo?.GetCustomAttributesData() ?? Array.Empty<CustomAttributeData>());
        attributes.AddRange(_underlyingProperty.GetCustomAttributesData());

        return attributes.AsReadOnly();
    }

    public override Type[] GetOptionalCustomModifiers()
        => _underlyingProperty.GetOptionalCustomModifiers();

    public override Type[] GetRequiredCustomModifiers()
        => _underlyingProperty.GetRequiredCustomModifiers();

    public override bool IsDefined(Type attributeType, bool inherit)
    {
        return (_constructionParameterInfo is not null && _constructionParameterInfo.IsDefined(attributeType, inherit)) ||
            _underlyingProperty.IsDefined(attributeType, inherit);
    }
}
```

`repro.csproj`:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  <PropertyGroup Condition="false">
      <UseMonoRuntime>true</UseMonoRuntime>
      <SelfContained>true</SelfContained>
  </PropertyGroup>
</Project>
```

**Expected output:**

```console
$ dotnet run
Got 2 attributes
FlargAttribute
BlargAttribute
```

**Actual output**

```console
$ dotnet run
Unhandled Exception:
System.ArgumentNullException: Value cannot be null. (Parameter 'attributeType')
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Type attributeType, Boolean inherit)
   at System.Reflection.RuntimePropertyInfo.GetCustomAttributes(Type attributeType, Boolean inherit)
   at Retrofit.PropertyAsParameterInfo.GetCustomAttributes(Type attributeType, Boolean inherit) in /private/tmp/parma/PropertyAsParameterInfo.cs:line 42
   at System.Reflection.CustomAttribute.GetCustomAttributesBase(ICustomAttributeProvider obj, Type attributeType, Boolean inheritedOnly)
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Type attributeType, Boolean inherit)
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Boolean inherit)
   at System.Attribute.GetCustomAttributes(ParameterInfo element)
   at System.Reflection.CustomAttributeExtensions.GetCustomAttributes(ParameterInfo element)
   at P.Main() in /private/tmp/parma/Program.cs:line 9
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentNullException: Value cannot be null. (Parameter 'attributeType')
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Type attributeType, Boolean inherit)
   at System.Reflection.RuntimePropertyInfo.GetCustomAttributes(Type attributeType, Boolean inherit)
   at Retrofit.PropertyAsParameterInfo.GetCustomAttributes(Type attributeType, Boolean inherit) in /private/tmp/parma/PropertyAsParameterInfo.cs:line 42
   at System.Reflection.CustomAttribute.GetCustomAttributesBase(ICustomAttributeProvider obj, Type attributeType, Boolean inheritedOnly)
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Type attributeType, Boolean inherit)
   at System.Reflection.CustomAttribute.GetCustomAttributes(ICustomAttributeProvider obj, Boolean inherit)
   at System.Attribute.GetCustomAttributes(ParameterInfo element)
   at System.Reflection.CustomAttributeExtensions.GetCustomAttributes(ParameterInfo element)
   at P.Main() in /private/tmp/parma/Program.cs:line 9
```

--- a/src/libraries/System.Runtime/tests/System.Reflection.Tests/CustomAttributeTests.cs
+++ b/src/libraries/System.Runtime/tests/System.Reflection.Tests/CustomAttributeTests.cs
@@ -129,5 +129,113 @@ namespace System.Reflection.Tests
             
             Assert.Same(string.Empty, attr.NamedField);
         }
+
+        [AttributeUsage(AttributeTargets.Parameter)]
+        internal class MyParameterAttribute : Attribute {}
+
+        [AttributeUsage(AttributeTargets.Property)]
+        internal class MyPropertyAttribute : Attribute {}
+
+        internal sealed class PropertyAsParameterInfo : ParameterInfo
+        {
+            private readonly PropertyInfo _underlyingProperty;
+            private readonly ParameterInfo? _constructionParameterInfo;
+
+            public PropertyAsParameterInfo(PropertyInfo property, ParameterInfo parameterInfo)
+            {
+                _underlyingProperty = property;
+                _constructionParameterInfo = parameterInfo;
+                MemberImpl = _underlyingProperty;
+            }
+
+            public override object[] GetCustomAttributes(Type attributeType, bool inherit)
+            {
+                var constructorAttributes = _constructionParameterInfo?.GetCustomAttributes(attributeType, inherit);
+
+                if (constructorAttributes == null || constructorAttributes is { Length: 0 })
+                {
+                    return _underlyingProperty.GetCustomAttributes(attributeType, inherit);
+                }
+
+                var propertyAttributes = _underlyingProperty.GetCustomAttributes(attributeType, inherit);
+
+                var mergedAttributes = new Attribute[constructorAttributes.Length + propertyAttributes.Length];
+                Array.Copy(constructorAttributes, mergedAttributes, constructorAttributes.Length);
+                Array.Copy(propertyAttributes, 0, mergedAttributes, constructorAttributes.Length, propertyAttributes.Length);
+
+                return mergedAttributes;
+            }
+
+            public override object[] GetCustomAttributes(bool inherit)
+            {
+                var constructorAttributes = _constructionParameterInfo?.GetCustomAttributes(inherit);
+
+                if (constructorAttributes == null || constructorAttributes is { Length: 0 })
+                {
+                    return _underlyingProperty.GetCustomAttributes(inherit);
+                }
+
+                var propertyAttributes = _underlyingProperty.GetCustomAttributes(inherit);
+
+                var mergedAttributes = new object[constructorAttributes.Length + propertyAttributes.Length];
+                Array.Copy(constructorAttributes, mergedAttributes, constructorAttributes.Length);
+                Array.Copy(propertyAttributes, 0, mergedAttributes, constructorAttributes.Length, propertyAttributes.Length);
+
+                return mergedAttributes;
+            }
+
+            public override IList<CustomAttributeData> GetCustomAttributesData()
+            {
+                var attributes = new List<CustomAttributeData>(
+                    _constructionParameterInfo?.GetCustomAttributesData() ?? Array.Empty<CustomAttributeData>());
+                attributes.AddRange(_underlyingProperty.GetCustomAttributesData());
+
+                return attributes.AsReadOnly();
+            }
+        }
+
+        internal class CustomAttributeProviderTestClass
+        {
+            public CustomAttributeProviderTestClass([MyParameter] int integerProperty)
+            {
+                IntegerProperty = integerProperty;
+            }
+
+            [MyProperty]
+            public int IntegerProperty { get; set; }
+        }
+
+        [Fact]
+        public void CustomAttributeProvider ()
+        {
+            var type = typeof(CustomAttributeProviderTestClass);
+            var propertyInfo = type.GetProperty(nameof(CustomAttributeProviderTestClass.IntegerProperty));
+            var ctorInfo = type.GetConstructor(new Type[] { typeof(int) });
+            var ctorParamInfo = ctorInfo.GetParameters()[0];
+            var propertyAndParamInfo = new PropertyAsParameterInfo(propertyInfo, ctorParamInfo);
+
+            // check GetCustomAttribute API
+            var cattrObjects = propertyAndParamInfo.GetCustomAttributes(true);
+            Assert.Equal(2, cattrObjects.Length);
+            Assert.Equal(typeof(MyParameterAttribute), cattrObjects[0].GetType());
+            Assert.Equal(typeof(MyPropertyAttribute), cattrObjects[1].GetType());
+
+            cattrObjects = propertyAndParamInfo.GetCustomAttributes(typeof(Attribute), true);
+            Assert.Equal(2, cattrObjects.Length);
+            Assert.Equal(typeof(MyParameterAttribute), cattrObjects[0].GetType());
+            Assert.Equal(typeof(MyPropertyAttribute), cattrObjects[1].GetType());
+
+            var cattrsEnumerable = propertyAndParamInfo.GetCustomAttributes();
+            Attribute[] cattrs = cattrsEnumerable.Cast<Attribute>().ToArray();
+            Assert.Equal(2, cattrs.Length);
+            Assert.Equal(typeof(MyParameterAttribute), cattrs[0].GetType());
+            Assert.Equal(typeof(MyPropertyAttribute), cattrs[1].GetType());
+
+            // check GetCustomAttributeData API
+            var customAttributesData = propertyAndParamInfo.GetCustomAttributesData();
+            Assert.Equal(2, customAttributesData.Count);
+            Assert.Equal(typeof(MyParameterAttribute), customAttributesData[0].AttributeType);
+            Assert.Equal(typeof(MyPropertyAttribute), customAttributesData[1].AttributeType);
+        }
     }
 }

--- a/src/mono/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs
+++ b/src/mono/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs
@@ -114,13 +114,7 @@ namespace System.Reflection
         // FIXME: Callers are explicitly passing in null for attributeType, but GetCustomAttributes prohibits null attributeType arguments
         internal static object[] GetCustomAttributesBase(ICustomAttributeProvider obj, Type? attributeType, bool inheritedOnly)
         {
-            object[] attrs;
-
-            if (IsUserCattrProvider(obj))
-                attrs = obj.GetCustomAttributes(attributeType!, true);
-            else
-                attrs = GetCustomAttributesInternal(obj, attributeType!, false);
-
+            object[] attrs = GetCustomAttributesInternal(obj, attributeType!, pseudoAttrs: false);
             //
             // All pseudo custom attributes are Inherited = false hence we can avoid
             // building attributes array which would be discarded by inherited checks
@@ -155,6 +149,9 @@ namespace System.Reflection
             if (!attributeType.IsSubclassOf(typeof(Attribute)) && !attributeType.IsInterface
                 && attributeType != typeof(Attribute) && attributeType != typeof(CustomAttribute) && attributeType != typeof(object))
                 throw new ArgumentException(SR.Argument_MustHaveAttributeBaseClass + " " + attributeType.FullName);
+
+            if (IsUserCattrProvider(obj))
+                return obj.GetCustomAttributes(attributeType, inherit);
 
             // FIXME: GetCustomAttributesBase doesn't like being passed a null attributeType
             if (attributeType == typeof(CustomAttribute))
@@ -305,6 +302,9 @@ namespace System.Reflection
         internal static object[] GetCustomAttributes(ICustomAttributeProvider obj, bool inherit)
         {
             ArgumentNullException.ThrowIfNull(obj);
+
+            if (IsUserCattrProvider(obj))
+                return obj.GetCustomAttributes(typeof(Attribute), inherit);
 
             if (!inherit)
                 return (object[])GetCustomAttributesBase(obj, null, false).Clone();

